vowpalwabbit/namespaced_feature_store.h

```diff
@@ -4,253 +4,217 @@
-    return lhs._indices - rhs._indices;
+    assert(lhs._indices_it >= rhs._indices_it);
+    return lhs._indices_it - rhs._indices_it;
   }
 
-  bool operator==(const indexed_iterator_t& rhs) const { return _indices == rhs._indices; }
-  bool operator!=(const indexed_iterator_t& rhs) const { return _indices != rhs._indices; }
+  bool operator==(const indexed_iterator_t& rhs) const { return _indices_it == rhs._indices_it; }
+  bool operator!=(const indexed_iterator_t& rhs) const { return _indices_it != rhs._indices_it; }
 };
 
 /// namespace_index - 1 byte namespace identifier. Either the first character of the namespace or a reserved namespace
 /// identifier namespace_hash - 8 byte hash
-struct namespaced_features
+struct namespaced_feature_store
 {
-  using iterator = iterator_t<features, namespace_index, uint64_t>;
-  using const_iterator = iterator_t<const features, const namespace_index, const uint64_t>;
-  using indexed_iterator = indexed_iterator_t<size_t, features, namespace_index, uint64_t>;
-  using const_indexed_iterator =
-      indexed_iterator_t<const size_t, const features, const namespace_index, const uint64_t>;
-
-  namespaced_features() = default;
-  ~namespaced_features() = default;
-  namespaced_features(const namespaced_features&) = default;
-  namespaced_features& operator=(const namespaced_features&) = default;
-  namespaced_features(namespaced_features&& other) = default;
-  namespaced_features& operator=(namespaced_features&& other) = default;
-
-  inline size_t size() const { return _feature_groups.size(); }
-  inline bool empty() const { return _feature_groups.empty(); }
+  using iterator = indexed_iterator_t<std::array<std::list<namespaced_features>, 256>,
+      std::vector<namespace_index>::iterator, std::list<namespaced_features>>;
+  using const_iterator = indexed_iterator_t<const std::array<std::list<namespaced_features>, 256>,
+      std::vector<namespace_index>::const_iterator, const std::list<namespaced_features>>;
+  using list_iterator = std::list<namespaced_features>::iterator;
+  using const_list_iterator = std::list<namespaced_features>::const_iterator;
+  using index_flat_iterator = VW::chained_proxy_iterator<list_iterator, features::audit_iterator>;
+
+  namespaced_feature_store() = default;
+  ~namespaced_feature_store() = default;
+
+  // For copies we deep copy the features but not the pool
+  namespaced_feature_store(const namespaced_feature_store& other)
+  {
+    _feature_groups = other._feature_groups;
+    _legacy_indices_existing = other._legacy_indices_existing;
+  }
+
+  namespaced_feature_store& operator=(const namespaced_feature_store& other)
+  {
+    _feature_groups = other._feature_groups;
+    _legacy_indices_existing = other._legacy_indices_existing;
+    return *this;
+  }
+  namespaced_feature_store(namespaced_feature_store&& other) = default;
+  namespaced_feature_store& operator=(namespaced_feature_store&& other) = default;
+
+  inline size_t size() const
```

Comment:
depending on how often we call `size()` is it worth re-calculating this or keeping the size ready every time something is added/removed?

Reply:
Yeah I wondered this, I don't thing its often and I wanted to avoid that state for now at least